### PR TITLE
Delete usage of FakeEtcdClient from master_test.go

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #15349

@timothysc

### CODE REVIEW

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please

File: pkg/master/master_test.go [hunk 1]
```diff
@@ -531,54 +527,62 @@ func testInstallThirdPartyAPIListVersion(t *testing.T, version string) {
...
+			if test.items != nil {
+				setupEtcdList(client, "/ThirdPartyResourceData/company.com/foos/default", test.items)
+			}
 
-		resp, err := http.Get(server.URL + "/apis/company.com/" + version + "/namespaces/default/foos")
-		if !assert.NoError(err) {
-			return
-		}
-		defer resp.Body.Close()
+			resp, err := http.Get(server.URL + "/apis/company.com/" + version + "/namespaces/default/foos")
```
Comment (timothysc) [inline pkg/master/master_test.go]: looks like gofmt or some weird tab insertion here the change appears unrelated.

File: pkg/master/master_test.go [hunk 2]
```diff
@@ -531,54 +527,62 @@ func testInstallThirdPartyAPIListVersion(t *testing.T, version string) {
...
-			if !reflect.DeepEqual(list.Items[ix], expectedObj) {
-				t.Errorf("expected:\n%#v\nsaw:\n%#v\n", expectedObj, list.Items[ix])
+			if len(list.Items) != len(test.items) {
+				t.Errorf("unexpected length: %d vs %d", len(list.Items), len(test.items))
+				return
 			}
-		}
+			// The order of elements in LIST is not guaranteed.
+			mapping := make(map[string]int)
+			for ix := range test.items {
```
Comment (timothysc) [inline pkg/master/master_test.go]: this change appears unrelated.

File: pkg/master/master_test.go [hunk 3]
```diff
@@ -603,20 +607,12 @@ func storeToEtcd(client tools.EtcdClient, path, name string, obj interface{}) er
 	return err
 }
 
-func setupEtcdList(fakeClient *tools.FakeEtcdClient, path string, list []Foo) error {
-	resp := tools.EtcdResponseWithError{
-		R: &etcd.Response{
-			Node: &etcd.Node{},
-		},
-	}
+func setupEtcdList(client tools.EtcdClient, path string, list []Foo) error {
```
Comment (timothysc) [inline pkg/master/master_test.go]: //TODO convert to storage here.

**Comment by timothysc** [comment]
So similar to other pr, is your intent to follow on to remove tools dependency?

File: pkg/master/master_test.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: I'm actually fixing the test here.
What I did is I opaqued the body of for loop into func() {}(), to make deferred functions to really be called at the end of the loop, not to have all of them be run at the end of the test.
So this is WAI.

File: pkg/master/master_test.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: Not really - before when we were using FakeEtcdClient - we were guaranteed the order of items in the returned list. This is no longer the case with real etcd - so we need to sort them before comparing them - there is a comment added above.

File: pkg/master/master_test.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: Done.

**Comment by wojtek-t** [comment]
>
So similar to other pr, is your intent to follow on to remove tools dependency?

Yes - I'm planning to have a separate PR for it.

**Comment by wojtek-t** [comment]
@timothysc - PTAL (I answered all your comments)

File: pkg/master/master_test.go [hunk 4]
```diff
@@ -531,54 +527,62 @@ func testInstallThirdPartyAPIListVersion(t *testing.T, version string) {
...
+			}
 
-		if test.items == nil {
-			if len(list.Items) != 0 {
-				t.Errorf("expected no items, saw: %v", list.Items)
+			if test.items == nil {
+				if len(list.Items) != 0 {
+					t.Errorf("expected no items, saw: %v", list.Items)
+				}
+				return
```
Comment (timothysc) [inline pkg/master/master_test.go]: Won't this exit the tests prematurely now?

File: pkg/master/master_test.go (hunk 2 above)
Comment (timothysc) [inline pkg/master/master_test.go]: k

File: pkg/master/master_test.go [hunk 5]
```diff
@@ -531,54 +527,62 @@ func testInstallThirdPartyAPIListVersion(t *testing.T, version string) {
...
+			for ix := range list.Items {
+				// Copy things that are set dynamically on the server
+				expectedObj := test.items[mapping[list.Items[ix].Name]]
+				expectedObj.SelfLink = list.Items[ix].SelfLink
+				expectedObj.ResourceVersion = list.Items[ix].ResourceVersion
+				expectedObj.Namespace = list.Items[ix].Namespace
+				expectedObj.UID = list.Items[ix].UID
+				expectedObj.CreationTimestamp = list.Items[ix].CreationTimestamp
+
+				if !reflect.DeepEqual(list.Items[ix], expectedObj) {
```
Comment (timothysc) [inline pkg/master/master_test.go]: Might want to denote that you're ensuring order so this function passes.

File: pkg/master/master_test.go [hunk 6]
```diff
@@ -804,10 +811,11 @@ func testInstallThirdPartyAPIDeleteVersion(t *testing.T, version string) {
 
 	assert.Equal(http.StatusNotFound, resp.StatusCode)
 
-	expectDeletedKeys := []string{etcdtest.PathPrefix() + "/ThirdPartyResourceData/company.com/foos/default/test"}
-
-	if !assert.True(reflect.DeepEqual(fakeClient.DeletedKeys, expectDeletedKeys)) {
-		t.Errorf("unexpected deleted keys: %v", fakeClient.DeletedKeys)
+	expectedDeletedKey := etcdtest.AddPrefix("ThirdPartyResourceData/company.com/foos/default/test")
+	_, err = client.Get(expectedDeletedKey, false, false)
+	etcdErr := err.(*etcd.EtcdError)
```
Comment (timothysc) [inline pkg/master/master_test.go]: Should probably use IsEtcdNotFound instead.

File: pkg/master/master_test.go [hunk 7]
```diff
@@ -886,12 +894,17 @@ func testInstallThirdPartyResourceRemove(t *testing.T, version string) {
...
+
+	expectedDeletedKeys := []string{
+		etcdtest.AddPrefix("/ThirdPartyResourceData/company.com/foos/default/test"),
+		etcdtest.AddPrefix("/ThirdPartyResourceData/company.com/foos/default/bar"),
 	}
-	if !assert.True(reflect.DeepEqual(fakeClient.DeletedKeys, expectDeletedKeys)) {
-		t.Errorf("unexpected deleted keys: %v", fakeClient.DeletedKeys)
+	for _, key := range expectedDeletedKeys {
+		_, err := client.Get(key, false, false)
+		etcdErr := err.(*etcd.EtcdError)
```
Comment (timothysc) [inline pkg/master/master_test.go]: same comment on error handling.

**Comment by timothysc** [comment]
@wojtek-t minor nits, then lgtm.

File: pkg/master/master_test.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: No - it's fine. It is just exiting the function, which is like "continue in a loop". So that works fine.

File: pkg/master/master_test.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: Done.

File: pkg/master/master_test.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: Done (good catch - thanks).

File: pkg/master/master_test.go (hunk 7 above)
Comment (wojtek-t) [inline pkg/master/master_test.go]: Done.

**Comment by wojtek-t** [comment]
@timothysc - thanks for the review!
Applying lgtm label based on your comment above.
